client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.spooling.encoding;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import io.trino.client.Column;
+import io.trino.client.spooling.encoding.JsonDecodingUtils.TypeDecoder;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.fasterxml.jackson.core.JsonToken.END_ARRAY;
+import static com.fasterxml.jackson.core.JsonToken.START_ARRAY;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterators.unmodifiableIterator;
+import static io.trino.client.spooling.encoding.JsonDecodingUtils.createTypeDecoders;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+
+public class JsonQueryDataAccess
+        implements QueryDataAccess
+{
+    private final InputStream stream;
+    private final TypeDecoder[] decoders;
+
+    public JsonQueryDataAccess(List<Column> columns, InputStream stream)
+    {
+        this.decoders = createTypeDecoders(columns);
```

Comment:
Is this creation necessary per instance? Seems like these are fixed for all segments and could be passed directly.

Reply:
This is WiP still

Reply:
Applied :)